### PR TITLE
[TT-211][TT-199] Mark recordings with their process type + allow multiple `OnRootFrameInit` calls

### PR DESCRIPTION
**Pull request by Domiii**
* paired w/ https://github.com/replayio/chromium-v8/pull/228
* https://linear.app/replay/issue/TT-211/mark-iframesextensionsdevtools-recordings-with-different-metadata

### CODE REVIEW

File: third_party/blink/renderer/bindings/core/v8/local_window_proxy.cc [hunk 1]
```diff
@@ -159,6 +160,20 @@ void LocalWindowProxy::DisposeContext(Lifecycle next_status,
   lifecycle_ = next_status;
 }
 
+static const char* RecordReplayGetProcessType(
+  bool isMainWorld, bool isRoot, bool isDevtools) {
```
Comment (toshok) [inline third_party/blink/renderer/bindings/core/v8/local_window_proxy.cc:164]: can this function be changed to take the World / Frame and do the checks here instead of passing 3 bools?e

**Review by toshok — COMMENTED** [review]
The process type stuff (modulo the 3 boolean parameter function) looks good, but it feels like allowing multiple `OnRootFrameInit` calls is going to have other implications that aren't addressed here:

* does it make sense to have `gReplayScriptsAlive` at all anymore?
* `gRootLocalFrame` also feels like something that we shouldn't have/rely on anymore (instead somehow getting a list of all root frames.)
* the default context stuff looks like it'll bite us again, right?

wdyt?

File: third_party/blink/renderer/bindings/core/v8/record_replay_interface.cc [hunk 2]
```diff
@@ -2672,13 +2679,6 @@ void OnRootFrameInit(v8::Isolate* isolate, LocalFrame* localFrame, v8::Local<v8:
       localFrame->IsCrossOriginToParentOrOuterDocument(),
       localFrame->GetDocument()->Url().GetString().Utf8().c_str()
       );
-
-  if (gReplayScriptsAlive) {
-    // Our "V8RecordReplaySetDefaultContext" logic implies a single local
-    // root frame per render process.
```
[the remark this answers is not in the document]
Comment (Domiii) [reply third_party/blink/renderer/bindings/core/v8/record_replay_interface.cc:2678]: It should be "active root frame".

**Comment by Domiii** [comment]
@toshok 

> * does it make sense to have `gReplayScriptsAlive` at all anymore?
Entirely different concept. Yes. You can have requests come in when replay scripts are not alive.

> * `gRootLocalFrame` also feels like something that we shouldn't have/rely on anymore, instead somehow getting a list of all root frames.  see `RecordReplayClearContexts` for a confluence of both `gRootLocalFrame`  / `gReplayScriptsAlive` and possibly now irrelevant code?
There is only one active local frame at a time.

> * the default context stuff looks like it'll bite us again, right?
It shouldn't 🤷‍♀️
If it does, we'll know! There are still unsolved mysteries regarding the wrong queries at the wrong time. Did not investigate that yet. That will happen next week.


TL;DR: The root frame change is a necessity, because else, simple navigation does not update the active root frame.

**Comment by toshok** [comment]
> @toshok
> 
> > * does it make sense to have `gReplayScriptsAlive` at all anymore?
> 
> Entirely different concept: You can still have requests come in when replay scripts are not alive.

But now that we're explicitly allowing multiple root frames to be initialized do we have 2 potential orderings?

1. command requests come in after existing root frame is being destroyed but before next root frame is initialized (this is essentially the same as what we have now)
2. command requests come in after after chrome has torn down the previous root frame "enough"/initialized the next root frame "enough" to switch the default context.  For some definition of "enough".

My concern is: can 2 happen?  if so, what happens then?  do we handle the requests in the new frame when they were meant for the old one?

> > * `gRootLocalFrame` also feels like something that we shouldn't have/rely on anymore, instead somehow getting a list of all root frames.  see `RecordReplayClearContexts` for a confluence of both `gRootLocalFrame`  / `gReplayScriptsAlive` and possibly now irrelevant code?

> There is only one active root frame at a time

Right, but my concern is we don't know if we're behaving properly wrt lifecycles of frames/contexts.  Certainly seems like we aren't given we can get requests when our script isn't alive.

> > * the default context stuff looks like it'll bite us again, right?
> 
> It shouldn't 🤷‍♀️ If it does, we'll know! There are still unsolved mysteries regarding the wrong queries at the wrong time. Did not investigate that yet. That will happen next week.

> TL;DR: The root frame change is a necessity, because else, simple navigation does not update the active root frame.

100% understand that 👍   I guess my concern is are we going trade navigation for different destabilization?

**Review by toshok — APPROVED** [review]

**Comment by Domiii** [comment]
>  we don't know if we're behaving properly wrt lifecycles of frames/contexts
We are definitely not "behaving properly" yet. Hence:
* https://linear.app/replay/issue/TT-204/[eval]-fix-framecontext-lifecycle-tracking-multiple-onrootframeinit-et
* https://linear.app/replay/issue/TT-200/investigate-eval-failures-part-5

**Comment by Domiii** [comment]
@toshok 
>  trade navigation for different destabilization in the meantime

There is no trade-off here. It does not matter what we do, its not correct yet. But this is definitely more correct than before, in that the command handler lives in the currently active window now, as opposed to before where the command handler lives in a dead frame.
